### PR TITLE
Add check-local to Github workflow

### PR DESCRIPTION
**Pull request by rbott**
This PR will add `make check-local` to the Github Actions workflow - hence add additional checks which uncover release-critical errors.

The NEWS file in the current master does not pass this test, but a fixed one is on its way with the `prepare-3.0` branch. Once that has been merged, I will rebase this PR.

Until then, reviews are welcome :-)

### CODE REVIEW

**Comment by rbott** [comment]
hmmm, I force-pushed a change to the workflow file but for some reason github does not pick up this change :-(
This PR is now doomed to fail every time, as there are leftovers from the unit tests which confuse check-local.

**Comment by rbott** [comment]
OK, now it fails as expected with the format of the NEWS file. That check should pass as soon as the `prepare-3.0` branch gets merged into master. We need to keep this open until then - but nevertheless please review :-)

File: Makefile.am [hunk 1]
```diff
@@ -2598,7 +2598,7 @@ ganeti:
 .PHONY: check-dirs
 check-dirs: $(GENERATED_FILES)
 	@set -e; \
-	find . -type d \( -name . -o -name .git -prune -o -print \) | { \
+	find . -type d -not -name '.' -not -path '*/\.*' | { \
```
Comment (saschalucas) [inline Makefile.am:2601]: Would it be sufficient to limit this to the root directory? i.e.
`find . -type d -not -name '.' -not -path './.*`
would be a more conservative approach. ATM I can not image what tool might clutter subdirs with .dot-folders? Just for curiosity: must `.` be escaped in `-path`?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -5,6 +5,8 @@ on: [push, pull_request]
 jobs:
   build:
     runs-on: ubuntu-latest
+    env:
+      LC_ALL: C
```
Comment (saschalucas) [inline .github/workflows/ci.yml:9]: Would it make sense to add this to Makefile.am? This way CI and manual invocation `make check-local` can be covered? AFAIK the check-news target is the only one, that requires LC_ALL=C. There are other parts in Makefile.am, where LC_ALL is set accordingly (search for LC_ALL).

**Comment by saschalucas** [comment]
currently CI fails: 
```bash
RELEASE=3.0.0~beta1 ./autotools/check-news < ./NEWS
Line 8: Unreleased version after current release 3.0.0 beta1
make: *** [Makefile:4863: check-news] Error 1
```
I assume this is expected and at the next release, incrementing the version in `configure.ac` is the last step we should do?

File: Makefile.am (hunk 1 above)
Comment (apoikos) [inline Makefile.am:2601]: @rbott, why is this change needed? The reason .git is ignored is because the build system knows how to filter it out from the final sources. If there's anything the build system does _not_ know how to handle, then I'd like it to fail in `check-local`.

**Comment by apoikos** [comment]
> OK, now it fails as expected with the format of the NEWS file. That check should pass as soon as the `prepare-3.0` branch gets merged into master. We need to keep this open until then - but nevertheless please review :-)

The breakage is expected (although the message is misleading), because we already bumped the version in `configure.ac` without setting a release date in `NEWS`. Under normal circumstances, this should not happen.

File: Makefile.am (hunk 1 above)
Comment (saschalucas) [reply Makefile.am:2601]: if you run `LC_ALL=C make check-local` it prints out:
```
Directory ./.github not listed in Makefile
Directory ./.github/workflows not listed in Makefile
make: *** [Makefile:4843: check-dirs] Error 1
```
The intention I assume, was to exclude `.dot-folders` at all (including .github). @apoikos: it sounds like this is the wrong way?

Comment (rbott) [reply Makefile.am:2601]: Thanks for the hin, that was probably a bit too much. `find . -type d -not -name '.' -not -path './.*'` works just as good and is easier to read :-) I will fix that!

Comment (saschalucas) [reply Makefile.am]: While the `check-local` target is also intended to be run locally by the developer and therefore it sounds good to exclude all `.dot-folders`, here it will no longer check for `.hpc` directory, which is enumerate in: https://github.com/ganeti/ganeti/blob/72fcdea9139c22201d6b19cc60a0525c2a345fe4/Makefile.am#L270 To get `check-local` into the CI, I suggest a minimal invasive way by adding .github only to the excludes dirs:
```make
check-dirs: $(GENERATED_FILES)
        @set -e; \
        find . -type d \( -name . -o -name .git -o - name .github -prune -o -print \) | { \
``` 
Probably we can see later, how to exclude more things locally.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (saschalucas) [reply .github/workflows/ci.yml:9]: looking deeper, it is the `check-news` target, that enforces the C locale: https://github.com/ganeti/ganeti/blob/72fcdea9139c22201d6b19cc60a0525c2a345fe4/autotools/check-news#L97-L101
Because `check-local` should run by CI and locally (includes `check-news`) I suggest to fix this in `Makefile.am`:
```make
check-news:
        LC_ALL=C RELEASE=$(PACKAGE_VERSION) $(CHECK_NEWS) < $(top_srcdir)/NEWS
```

**Comment by rbott** [comment]
I am not entirely sure this PR / `make check-local` (or more specific: the versioning/NEWS part) will add or resolve more problems. So maybe we should drop the commit which adds it to the Github Actions workflow?

File: Makefile.am (hunk 1 above)
Comment (rbott) [reply Makefile.am]: Good catch, I'll update that accordingly!

Comment (rbott) [reply Makefile.am]: I have updated this PR to only exclude .git and .github folders. This way we do not falsly exclude needed folders and also have a clean list of what is supposed to be excluded.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (rbott) [reply .github/workflows/ci.yml:9]: I was wondering: could there be any other parts of the build process which actually rely on LC_ALL=C? Just setting it for a specific build step might break things at a later point.

File: Makefile.am (hunk 1 above)
Comment (rbott) [reply Makefile.am]: This has been fixed with the latest commit.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (iustin) [reply .github/workflows/ci.yml:9]: I think it would be a good idea to make entire build work work an UTF-8 locale, even if C.UTF-8. Plain C seems so archaic :)

Comment (rbott) [reply .github/workflows/ci.yml:9]: Sounds reasonable to me. But I would propose a separate PR for this. I suppose this should receive some through testing (especially regarding documentation). In any case I'd take a shot at it.

Comment (iustin) [reply .github/workflows/ci.yml:9]: Oh, definitely didn't mean in this PR. Was just surprised seeing we rely on C...

**Comment by rbott** [comment]
Pinging @saschalucas @iustin @apoikos: what is your current take on this PR?

**Comment by saschalucas** [comment]
> Pinging @saschalucas @iustin @apoikos: what is your current take on this PR?

I'm ok with this PR, if the locale setting gets moved from `.github/workflows/ci.yml` to `Makefile.am`, so that `check-local` can be run manually, too. Later we should investigate why different parts of the code need different locales and move it to a single common locale.

**Comment by rbott** [comment]
Interestingly, @apoikos implemented C (or rather: C.UTF-8) partially here in the Makefile: https://github.com/ganeti/ganeti/blob/4d8c16a0739c93400471023b3cf9adf73a037b8f/Makefile.am#L35
Related commit is 8563fdf57574e14d1a0da0727d11622dbc9f9145

The comment states `Use C.UTF-8 for tests/checks when available, fall back to en_US.UTF-8 else`, however the locale gets only set/used in the context of man pages and API docs. I would suggest to use this as a general setting for the whole build process. Does this seem right to everyone?

**Comment by rbott** [comment]
I will close this PR for now (but take a look at the locale thing discussed above and possibly open a new PR for that).
